### DISCUSSION TITLE
Windows ipp-over-usb

### DISCUSSION THREAD

**Discussion by luc-git** [alexpevzner/sane-airscan#318] [Q&A]
Hello does windows `IPP Class Driver` supports ipp over usb scanners?

**Comment by alexpevzner** [comment]
Honestly, I don't know.

**Reply by luc-git** [reply]
https://learn.microsoft.com/en-us/windows-hardware/drivers/print/end-of-servicing-plan-for-third-party-printer-drivers-on-windows?WT.mc_id=AZ-MVP-5004580

```
Q: Will multi-function devices (print, scan, fax) work over IPP?

A: Yes, they will.  For network devices, the Print and Fax endpoints will work via IPP and IPP Fax Out, respectively, while the Scan endpoint will work via WS-Scan or eSCL.  For USB devices, the endpoints will only be accessible when the USB interface is in IPP Over USB mode which means IPP for Print, IPP Fax Out for Fax, and only eSCL for Scan.
```

Ok so if I correctly understood this source tells that scanner over ipp over usb scanner is or will be supported.

**Comment by alexpevzner** [comment]
I'd better test rather to blindly trust that article. That article may explain what was planned a whole ago, not what was actually implemented and released.

USB traffic on Windows can be captured using USBCap (https://desowin.org/usbpcap/) or quite recent version of WireShark. So if scanner does use eSCL over IPP over USB, it can easily be tested.

What is interesting, if Windows supports IPP over USB, does it expose it as pseudo-network service on `localhost`, as Linux does, or some kind of non-network transport for the IPP/eSCL drivers working with local devices is provided.

If IPP over USB devices are exposed as pseudo-network devices, how to discover TCP port being used? On Linux it is announced via DNS-SD, using Avahi, as for normal network devices

**Reply by luc-git** [reply]
But if you haven't a scanner capable of escl it's difficult. Mine is a scx-3400 series and not escl compatible

**Reply by luc-git** [reply]
https://support.microsoft.com/en-us/topic/june-28-2022-kb5014666-os-builds-19042-1806-19043-1806-and-19044-1806-preview-4bd911df-f290-4753-bdec-a83bc8709eb6

> IPP over USB support – Microsoft has supported Internet Print Protocol (IPP) for network printers starting in 2018 with the release of Windows 10, version 1809. We are now expanding IPP support to USB printers.

Another article found but maybe it's not sufficient but we never know
